clang/lib/AST/ASTContext.cpp

```diff
@@ -1687,7 +1687,8 @@ CharUnits ASTContext::getDeclAlign(const Decl *D, bool ForAlignof) const {
       if (VD->hasGlobalStorage() && !ForAlignof) {
         uint64_t TypeSize =
             !BaseT->isIncompleteType() ? getTypeSize(T.getTypePtr()) : 0;
-        Align = std::max(Align, getTargetInfo().getMinGlobalAlign(TypeSize));
+        Align = std::max(Align, getTargetInfo().getMinGlobalAlign(
+                                    TypeSize, VD->hasDefinition()));
```

Comment:
Not sure hasDefinition() is quite the right thing here; what if the symbol is weak?

Reply:
Good point!

It seems to me though that this is something the backend must handle. Even with this patch, e.g. a weak 4-byte integer would get the natural alignment, and could later be replaced with an unaligned 4-byte integer by the linker. This patch only removes the ABI minimal alignment, i.e. the requirement to align everything by at least 2 bytes.

A recent GCC does not handle this either: it emits a LARL for a weak int. Could it be that the linker preserves the alignment of the original (weak) object?

Reply:
So the intent of the option is to 1) when emitting a symbol definition directly, keep using the 2-byte ABI alignment, but 2) when refering to a symbol that *might* be defined externally, do not make using any alignment assumption.  In GCC this is using the same checks used elsewhere to test where a symbol is guaranteed local or might be external (keeping in mind things like weak symbols or ELF overrides).

Reply:
I can't see that this is working with weak symbols with (recent) GCC:

```
test2.c:
 #include <stdio.h> 

char __attribute__((weak)) c0_weak = 0;
char __attribute__((weak)) c1_weak = 0;
 
int main() { 
  printf("%d\n", c0_weak + c1_weak); 
}

test2_1.c:
char __attribute__((aligned(1))) c0_weak = 1;
char __attribute__((aligned(1))) c1_weak = 1;
```

```
gcc ./test2.c ./test2_1.c -O3 -munaligned-symbols -S -c
main:
.LFB11:
	.cfi_startproc
	larl	%r2,c0_weak
	larl	%r1,c1_weak
        ...
```

```
~/gcc-latest/bin/gcc ./test2.c ./test2_1.c -O3 -munaligned-symbols
/usr/bin/ld: /tmp/ccj8BnuO.o(.text.startup+0x2): misaligned symbol `c0_weak' (0x1004021) for relocation R_390_PC32DBL
collect2: error: ld returned 1 exit status
```

Reply:
Hi. I've just committed a fix for GCC to handle weak symbols correctly with -munaligned-symbols:
https://gcc.gnu.org/pipermail/gcc-patches/2023-December/640893.html